src/httpserver.cpp

```diff
@@ -940,6 +941,92 @@ bool HTTPRequest::LoadBody(LineReader& reader)
+{
+    HTTPResponse res;
+
+    // Some response headers are determined in advance and stored in the request
+    res.m_headers = std::move(m_response_headers);
+
+    // Response version matches request version
+    res.m_version_major = m_version_major;
+    res.m_version_minor = m_version_minor;
+
+    // Add response code and look up reason string
+    res.m_status = status;
+    res.m_reason = HTTPReason.find(status)->second;
+
+    // See libevent evhttp_response_needs_body()
+    // Response headers are different if no body is needed
+    bool needs_body{status != HTTP_NO_CONTENT && (status < 100 || status >= 200)};
+
+    // See libevent evhttp_make_header_response()
+    // Expected response headers depend on protocol version
+    if (m_version_major == 1) {
+        // HTTP/1.0
+        if (m_version_minor == 0) {
+            auto connection_header{m_headers.Find("Connection")};
+            if (connection_header && ToLower(connection_header.value()) == "keep-alive") {
+                res.m_headers.Write("Connection", "keep-alive");
+                res.m_keep_alive = true;
+            }
+        }
+
+        // HTTP/1.1
+        if (m_version_minor >= 1) {
+            const int64_t now_seconds{TicksSinceEpoch<std::chrono::seconds>(NodeClock::now())};
+            res.m_headers.Write("Date", FormatRFC7231DateTime(now_seconds));
+
+            if (needs_body) {
+                res.m_headers.Write("Content-Length", strprintf("%d", reply_body.size()));
+            }
+
+            // Default for HTTP/1.1
+            res.m_keep_alive = true;
+        }
+    }
+
+    if (needs_body && !res.m_headers.Find("Content-Type")) {
+        // Default type from libevent evhttp_new_object()
+        res.m_headers.Write("Content-Type", "text/html; charset=ISO-8859-1");
+    }
+
+    auto connection_header{m_headers.Find("Connection")};
+    if (connection_header && ToLower(connection_header.value()) == "close") {
+        // Might not exist already but we need to replace it, not append to it
+        res.m_headers.Remove("Connection");
+        res.m_headers.Write("Connection", "close");
+        res.m_keep_alive = false;
+    }
+
+    // Serialize the response headers
+    const std::string headers{res.StringifyHeaders()};
+    const auto headers_bytes{std::as_bytes(std::span(headers.begin(), headers.end()))};
+
+    // Fill the send buffer with the complete serialized response headers + body
+    {
+        LOCK(m_client->m_send_mutex);
+        m_client->m_send_buffer.insert(m_client->m_send_buffer.end(), headers_bytes.begin(), headers_bytes.end());
+
+        // We've been using std::span up until now but it is finally time to copy
+        // data. The original data will go out of scope when WriteReply() returns.
+        // This is analogous to the memcpy() in libevent's evbuffer_add()
+        m_client->m_send_buffer.insert(m_client->m_send_buffer.end(), reply_body.begin(), reply_body.end());
```

Comment:
Maybe use a list of buffers (instead a single FIFO buffer)?

Reply:
I suggest considering a similar approach to `libevent`'s evbuffers:

> evbuffers are represented using a linked list of memory chunks, with pointers to the first and last chunk in the chain.

https://libevent.org/doc/buffer_8h.html

Reply:
Thinking about it, doing the copy here is definitely fine for now (we can optimize it in a following PR).